### PR TITLE
[Core] manage nccl via a pypi package & upgrade to pt 2.2.1

### PR DESCRIPTION
**Pull request by youkaichao**
See https://github.com/vllm-project/vllm/pull/3802 for why we don't use another pypi packge to manage nccl .

See https://github.com/NVIDIA/nccl/issues/1234 on why we have to manage nccl when upgrading to pt 2.2.1 .

Usage:

```python
python -m vllm.tools.install_nccl --cuda 12 --nccl 2.18.3
```

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -49,7 +49,7 @@ jobs:
       matrix:
           os: ['ubuntu-20.04']
           python-version: ['3.8', '3.9', '3.10', '3.11']
-          pytorch-version: ['2.1.2']  # Must be the most recent version that meets requirements.txt.
+          pytorch-version: ['2.2.1']  # Must be the most recent version that meets requirements.txt.
```
Comment (WoosukKwon) [inline .github/workflows/publish.yml:52]: Can we upgrade torch 2.2.2 instead?

File: requirements.txt [hunk 2]
```diff
@@ -17,3 +17,4 @@ pynvml == 11.5.0
 triton >= 2.1.0
 outlines == 0.0.34
 tiktoken == 0.6.0 # Required for DBRX tokenizer
+vllm-nccl-cu12>=2.18<2.19 # for downloading nccl library
```
Comment (WoosukKwon) [inline requirements.txt:20]: Can we do this?
```suggestion
vllm-nccl-cu12>=2.18.3<2.19 # for downloading nccl library
```

File: setup.py [hunk 3]
```diff
@@ -328,6 +328,15 @@ def get_requirements() -> List[str]:
...
             requirements = f.read().strip().split("\n")
+            cuda_major = torch.version.cuda.split(".")[0]
+            modified_requirements = []
+            for req in requirements:
+                if "vllm-nccl-cu12" in req:
+                    modified_requirements.append(
+                        req.replace("vllm-nccl-cu12",
+                                    f"vllm-nccl-cu{cuda_major}"))
+                else:
+                    modified_requirements.append(req)
```
Comment (WoosukKwon) [inline setup.py:339]: Shouldn't we add `requirements = modified_requirements`?

File: vllm/model_executor/parallel_utils/pynccl.py [hunk 4]
```diff
@@ -40,12 +41,17 @@
         f"Loading nccl from environment variable VLLM_NCCL_SO_PATH={so_file}")
 else:
     if torch.version.cuda is not None:
-        so_file = "libnccl.so.2"
+        cuda_major = torch.version.cuda.split(".")[0]
+        # check if we have vllm-managed nccl
+        path = os.path.expanduser(
+            f"~/.config/vllm/nccl/cu{cuda_major}/libnccl.so.*")
```
Comment (WoosukKwon) [inline vllm/model_executor/parallel_utils/pynccl.py:47]: nit: Let's define this path as a global constant variable?

**Review by WoosukKwon — COMMENTED** [review]
@youkaichao Awesome! Thanks for the great work! ! I didn't realize that updating PyTorch could be this challenging. Left minor comments. Please check them out.

File: .github/workflows/publish.yml (hunk 1 above)
Comment (youkaichao) [reply .github/workflows/publish.yml:52]: Technically we can, but I'm afraid it's too new and has not been verified by the community yet (e.g. maybe some new bugs). Let's wait for some time (say one month) to decide if we want to move to pytorch 2.2.2 ?

File: requirements.txt (hunk 2 above)
Comment (youkaichao) [reply requirements.txt:20]: Fun fact: actually I have https://pypi.org/project/vllm-nccl-cu12/2.18.1.0.1.0/ , which is nccl 2.18.1 . I don't find nccl 2.18.3 at https://developer.download.nvidia.com/compute/redist/nccl/ .

File: .github/workflows/publish.yml (hunk 1 above)
Comment (WoosukKwon) [reply .github/workflows/publish.yml:52]: I see. Makes sense!

File: setup.py (hunk 3 above)
Comment (youkaichao) [reply setup.py:339]: Nice catch! Fixed in [fee6165](https://github.com/vllm-project/vllm/pull/3805/commits/fee61656e65cd8ce5dc5ce959f86782a71f7071b)

File: vllm/model_executor/parallel_utils/pynccl.py (hunk 4 above)
Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py]: fixed in [21bffe9](https://github.com/vllm-project/vllm/pull/3805/commits/21bffe9b4540088910efb9ab817dbda70ef90a1e)

File: vllm/model_executor/parallel_utils/pynccl.py [hunk 5]
```diff
@@ -34,18 +35,24 @@
 
 so_file = os.environ.get("VLLM_NCCL_SO_PATH", "")
 
+# check if we have vllm-managed nccl
+cuda_major = torch.version.cuda.split(".")[0]
```
Comment (WoosukKwon) [inline vllm/model_executor/parallel_utils/pynccl.py:39]: QQ: Does this work with AMD?

Comment (youkaichao) [reply vllm/model_executor/parallel_utils/pynccl.py]: Emm, good question. Fixed in [bbf6afe](https://github.com/vllm-project/vllm/pull/3805/commits/bbf6afe2ba0e102931e44669b0764e6a9f149ca1).

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks again for the great work!

**Pull request merged**